src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,129 @@
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+
+		string archiveName = GetOSFileName();
+		string? downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		long backupPathTimeStamp = DateTime.Now.Millisecond;
+		string backupPath = Path.Join( BmxPaths.OLD_BMX_VERSIONS_PATH, $"bmx-v{localVersion}-{backupPathTimeStamp}-old.bak" );
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			throw new BmxException( "Could move the old version, try with elevated permissions", ex );
+		} catch {
+			throw new BmxException( "BMX could not update" );
+		}
+
+
+		string downloadPath = Path.GetTempFileName();
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None );
+			await archiveRes.Content.CopyToAsync( fs );
+			await fs.FlushAsync();
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
+			throw new BmxException( "Failed to download the update", ex );
+		}
+
+		try {
+			string extension = Path.GetExtension( downloadUrl );
+
+			if( extension.Equals( ".zip", StringComparison.OrdinalIgnoreCase ) ) {
+				ExtractZipFile( downloadPath, currentDirectory );
+			} else if( extension.Equals( ".gz", StringComparison.OrdinalIgnoreCase ) ) {
+				ExtractTarGzipFile( downloadPath, currentDirectory );
+			} else {
+				throw new Exception( "Unknown archive type" );
+			}
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
+			throw new BmxException( "Failed to update with new files", ex );
+		} finally {
+			File.Delete( downloadPath );
+		}
+	}
+
+	private static string GetOSFileName() {
+
+		if( RuntimeInformation.IsOSPlatform( OSPlatform.Windows ) ) {
+			return "bmx-win-x64.zip";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.OSX ) ) {
+			return "bmx-osx-x64.tar.gz";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.Linux ) ) {
+			return "bmx-linux-x64.tar.gz";
+		} else {
+			throw new BmxException( "New version does not support you current OS" );
+		}
+	}
+
+	public static void Cleanup() {
+		if( !Directory.Exists( BmxPaths.OLD_BMX_VERSIONS_PATH ) ) {
+			try {
+				Directory.CreateDirectory( BmxPaths.OLD_BMX_VERSIONS_PATH );
+			} catch( Exception ex ) {
+				throw new BmxException( "Failed to initialize temporary BMX file directory (~/.bmx/temp)", ex );
+			}
+		}
```

Comment:
I don't like that we're creating a directory in a function named "Cleanup", and that our update logic relies on this cleanup function running first:
https://github.com/Brightspace/bmx/pull/418/files#diff-48acde0deb16dbeda7f36eb7087c301cf5a486a4fe2044fc77522fdd67191353R37-R39

Also not a fan of always creating the temp directory even when it's usually not needed, but this isn't a deal breaker. If we do want to always ensure the existence of this temp directory, we should do this together with other directory init logic here:
https://github.com/Brightspace/bmx/blob/8c28c21d264da16e15b8af636b58e14dde192cf5/src/D2L.Bmx/Program.cs#L215-L227

Reply:
Gotcha! How about creating a new directory every update call and then deleting the directory if it exists in the clean up so the directory itself doesn't exist when it is not needed? Would that be an overkill of creation and deletions of directories?

Reply:
I think that makes sense. easier cleanup should be one of the benefits of using a subdirectory